From d11f622a810f305508c436ff70ff07bb437f8eec Mon Sep 17 00:00:00 2001
From: CI Assistant <yossi.ilkanaev@intel.com>
Date: Mon, 1 Oct 2018 15:11:27 +0200
Subject: [PATCH] Merge pull request #190 in SW_UGW/linux from
 feature/UGW_SW-30344-linux-firewall-rule-update-for-sdl to master

* commit '6c05abda5dbf928d7761ec95d0f2e7a075163b78':
  UGW_SW-30344: IPTable event handling kernel patching or event generation part
---
 include/net/ip.h               |  2 ++
 include/net/ppa/ppa_stack_al.h | 10 ++++++++
 net/ipv4/netfilter/ip_tables.c | 58 ++++++++++++++++++++++++++++++++++++++++++
 net/netfilter/Kconfig          |  7 +++++
 4 files changed, 77 insertions(+)

diff --git a/include/net/ip.h b/include/net/ip.h
index bc9b4deeb60e..12640ec08159 100644
--- a/include/net/ip.h
+++ b/include/net/ip.h
@@ -98,6 +98,8 @@ extern struct ip_ra_chain __rcu *ip_ra_chain;
 
 #define IP_FRAG_TIME	(30 * HZ)		/* fragment lifetime	*/
 
+#define IPTABLE_CHANGE 0x0005		/* PPA iptable change for SDL	*/
+
 struct msghdr;
 struct net_device;
 struct packet_type;
diff --git a/include/net/ppa/ppa_stack_al.h b/include/net/ppa/ppa_stack_al.h
index 7ae2e8b2d944..520f5b9589aa 100644
--- a/include/net/ppa/ppa_stack_al.h
+++ b/include/net/ppa/ppa_stack_al.h
@@ -607,8 +607,18 @@ typedef struct list_head		PPA_LIST_NODE;
 /*!
 	\brief netdevice event registeration/unregisteration functions
  */
+extern int register_iptable_notifier(struct notifier_block *nb);
+extern int unregister_iptable_notifier(struct notifier_block *nb);
+
+#define ppa_register_iptable_notifier register_iptable_notifier
+#define ppa_unregister_iptable_notifier unregister_iptable_notifier
+
+/*
+	This is for iptable notifier, not for netdev notification
+*/
 #define ppa_register_netdevice_notifier register_netdevice_notifier
 #define ppa_unregister_netdevice_notifier unregister_netdevice_notifier
+
 /*
  * ####################################
  * Inline Functions
diff --git a/net/ipv4/netfilter/ip_tables.c b/net/ipv4/netfilter/ip_tables.c
index e78f6521823f..fa135ee5945b 100644
--- a/net/ipv4/netfilter/ip_tables.c
+++ b/net/ipv4/netfilter/ip_tables.c
@@ -1122,6 +1122,58 @@ __do_replace(struct net *net, const char *name, unsigned int valid_hooks,
 	return ret;
 }
 
+#if IS_ENABLED(CONFIG_PPA_IPTABLE_EVENT_HANDLING)
+static RAW_NOTIFIER_HEAD(iptable_chain);
+
+static int call_iptable_notifier(struct notifier_block *nb, unsigned long val)
+{
+	return nb->notifier_call(nb, val, NULL);
+}
+
+int register_iptable_notifier(struct notifier_block *nb)
+{
+	int err;
+
+	rtnl_lock();
+	err = raw_notifier_chain_register(&iptable_chain, nb);
+	if (err)
+		goto unlock;
+
+	err = call_iptable_notifier(nb, IPTABLE_CHANGE);
+	err = notifier_to_errno(err);
+	if (err)
+		raw_notifier_chain_unregister(&iptable_chain, nb);
+
+unlock:
+	rtnl_unlock();
+	return err;
+}
+EXPORT_SYMBOL(register_iptable_notifier);
+
+int unregister_iptable_notifier(struct notifier_block *nb)
+{
+	int err;
+
+	rtnl_lock();
+	err = raw_notifier_chain_unregister(&iptable_chain, nb);
+	rtnl_unlock();
+
+	return err;
+}
+EXPORT_SYMBOL(unregister_iptable_notifier);
+
+static int call_iptable_notifiers_info(unsigned long val)
+{
+	return raw_notifier_call_chain(&iptable_chain, val, NULL);
+}
+
+int call_iptable_notifiers(void)
+{
+	return call_iptable_notifiers_info(1);
+}
+EXPORT_SYMBOL(call_iptable_notifiers);
+#endif
+
 static int
 do_replace(struct net *net, const void __user *user, unsigned int len)
 {
@@ -1690,6 +1742,12 @@ do_ipt_set_ctl(struct sock *sk, int cmd, void __user *user, unsigned int len)
 	switch (cmd) {
 	case IPT_SO_SET_REPLACE:
 		ret = do_replace(sock_net(sk), user, len);
+#if IS_ENABLED(CONFIG_PPA_IPTABLE_EVENT_HANDLING)
+		call_iptable_notifiers();
+		/* invokes the ppa handler for flushing the sessions
+		programmed currently in the HW on any modifications
+		made on iptable rules. */
+#endif
 		break;
 
 	case IPT_SO_SET_ADD_COUNTERS:
diff --git a/net/netfilter/Kconfig b/net/netfilter/Kconfig
index 330613bebedc..79030e5eb2be 100644
--- a/net/netfilter/Kconfig
+++ b/net/netfilter/Kconfig
@@ -969,6 +969,13 @@ config NETFILTER_XT_TARGET_SECMARK
 
 	  To compile it as a module, choose M here.  If unsure, say N.
 
+config PPA_IPTABLE_EVENT_HANDLING
+	tristate 'PPA iptable event handling'
+	default y
+	help
+	  SDL Firewall Rule update for iptable with respect to PPA
+	  sessions programmed into the hardware
+
 config NETFILTER_XT_TARGET_TCPMSS
 	tristate '"TCPMSS" target support'
 	depends on IPV6 || IPV6=n
